### PR TITLE
Add ActivityScheduleTimeout deduction logic

### LINKED ISSUE #589: Do not require heartbeat timeout
It is surprising that heartbeat timeout is required:
https://github.com/uber/cadence/blob/master/service/history/historyEngine.go#L1903

Also I would not require other timeouts. My proposal is to require either:

ScheduleToClose or both ScheduleToStart and StartToClose.

### PR DESCRIPTION
**Pull request by arthurguber**
This commit relaxes validateActivityScheduleAttributes by having it fill in some
of the fields related to schedule timeouts if they are empty.

The logic used is identical to the logic used in AWS SWF's java client.

The logic is as follows:
If ScheduleToClose is present and ScheduleToStart or StartToClose is missing,
we will set what is missing to ScheduleToClose.

If ScheduleToClose is missing but both ScheduleToStart and StartToClose are present,
we will set ScheduleToClose to the sum of ScheduleToStart and StartToClose.

resolves #589

### CODE REVIEW

File: service/history/historyEngine.go [hunk 1]
```diff
@@ -2397,6 +2388,23 @@ func validateActivityScheduleAttributes(attributes *workflow.ScheduleActivityTas
 		}
 	}
 
+	validScheduleToClose := attributes.GetScheduleToCloseTimeoutSeconds() > 0
```
Comment (yiminc) [inline service/history/historyEngine.go]: we need to error out when value < 0

File: service/history/historyEngine.go [hunk 2]
```diff
@@ -2397,6 +2388,23 @@ func validateActivityScheduleAttributes(attributes *workflow.ScheduleActivityTas
...
+	validScheduleToStart := attributes.GetScheduleToStartTimeoutSeconds() > 0
+	validStartToClose := attributes.GetStartToCloseTimeoutSeconds() > 0
+
+	if validScheduleToClose {
+		if !validScheduleToStart {
```
Comment (yiminc) [inline service/history/historyEngine.go]: we should do the overwrite only when the ScheduleToStartTimeoutSeconds is 0

File: service/history/historyEngine.go [hunk 3]
```diff
@@ -2397,6 +2388,23 @@ func validateActivityScheduleAttributes(attributes *workflow.ScheduleActivityTas
...
+			attributes.ScheduleToStartTimeoutSeconds = common.Int32Ptr(attributes.GetScheduleToCloseTimeoutSeconds())
+		}
+		if !validStartToClose {
+			attributes.StartToCloseTimeoutSeconds = common.Int32Ptr(attributes.GetScheduleToCloseTimeoutSeconds())
```
Comment (yiminc) [inline service/history/historyEngine.go]: same as above

File: service/history/historyEngine.go [hunk 4]
```diff
@@ -2397,6 +2388,23 @@ func validateActivityScheduleAttributes(attributes *workflow.ScheduleActivityTas
...
+		}
+	} else if validScheduleToStart && validStartToClose {
+		attributes.ScheduleToCloseTimeoutSeconds = common.Int32Ptr(attributes.GetScheduleToStartTimeoutSeconds() + attributes.GetStartToCloseTimeoutSeconds())
```
Comment (yiminc) [inline service/history/historyEngine.go]: and same here, only overwrite when ScheduleToCloseTimeoutSeconds is 0

File: service/history/historyEngine.go (hunk 2 above)
Comment (arthurguber) [reply service/history/historyEngine.go]: resolved

File: service/history/historyEngine.go (hunk 3 above)
Comment (arthurguber) [reply service/history/historyEngine.go]: resolved

File: service/history/historyEngine.go (hunk 4 above)
Comment (arthurguber) [reply service/history/historyEngine.go]: resolved

File: service/history/historyEngine.go [hunk 5]
```diff
@@ -2373,15 +2373,6 @@ func validateActivityScheduleAttributes(attributes *workflow.ScheduleActivityTas
...
-		return &workflow.BadRequestError{Message: "A valid StartToCloseTimeoutSeconds is not set on decision."}
-	}
-	if attributes.ScheduleToStartTimeoutSeconds == nil || *attributes.ScheduleToStartTimeoutSeconds <= 0 {
-		return &workflow.BadRequestError{Message: "A valid ScheduleToStartTimeoutSeconds is not set on decision."}
-	}
-	if attributes.ScheduleToCloseTimeoutSeconds == nil || *attributes.ScheduleToCloseTimeoutSeconds <= 0 {
-		return &workflow.BadRequestError{Message: "A valid ScheduleToCloseTimeoutSeconds is not set on decision."}
-	}
 	if attributes.HeartbeatTimeoutSeconds == nil || *attributes.HeartbeatTimeoutSeconds < 0 {
 		return &workflow.BadRequestError{Message: "A valid HeartbeatTimeoutSeconds is not set on decision."}
```
Comment (mfateev) [inline service/history/historyEngine.go]: Why heartbeat is required?

Comment (arthurguber) [reply service/history/historyEngine.go]: resolved

**Pull request merged**